### PR TITLE
Add link checking step to CI pipeline

### PR DESCRIPTION
**Pull request by niharikasingh**
I've made the link-checking step a non-blocking step. If any links are failing, the pipeline will add a comment to the PR but will otherwise succeed. There are no real failing links in this repo, but I've left the failing comment here so you can see what it looks like. In real life, we would just delete this after resolving all errors. 

Explanation of commands used: 
- `find public -name "*.js" -exec grep -Eo "(http|https)://[a-zA-Z0-9./?=_%:-]+" {} \; | sort -u > linklist.txt`: Find all built javascript files, grep for all URLs using the given regex, sort and deduplicate those URLs, and store them in `linklist.txt`
- `printf '%s\n%s\n' "# LinkChecker URL list" "$(cat linklist.txt)" > linklist.txt`: Add a comment to the top of the output file so that the linkchecker works 
- `linkchecker linklist.txt --check-extern --ignore-url="https://.*\.fastly\.net/.*" -o failures > output.txt || true`: Use the link checker to check each URL in the output file. Ignore the given regexes (we can add to this to ignore more files). Put all failing links in `output.txt`. Return a successful exit code no matter what because we want this pipeline to succeed. 
- `cat output.txt`: Print all failing links so that we can look at them later. 
- `echo $(wc -l < output.txt) >> $GITHUB_OUTPUT`: Store the number of failing links so we can put it in a PR comment. 

List of links being checked: 
[linklist.txt](https://github.com/georgetown-cset/eto-supply-chain/files/12613025/linklist.txt)

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -22,15 +22,15 @@ jobs:
       run: |
         python -m pip install --upgrade pip
         pip install -r requirements.txt
-    - name: Test with pytest
```
Comment (jmelot) [inline .github/workflows/main.yml:25]: Some of these changes look unrelated, should they be reverted?

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -54,19 +54,39 @@ jobs:
...
+        printf '%s\n%s\n' "# LinkChecker URL list\n# <meta charset="UTF-8">" "$(cat linklist.txt)" > linklist.txt
+        linkchecker linklist.txt --check-extern --ignore-url="https://.*\.fastly\.net/.*" --ignore-url="https://.*\.mapbox\..*" -o failures > output.txt || true
+        cat output.txt
+        echo "link_check=$(wc -l < output.txt | sed 's/^ *//g')" >> $GITHUB_OUTPUT
+    - name: Edit PR comment about link checking
+      if: steps.link_check.outputs.link_check > 0
+      uses: thollander/actions-comment-pull-request@v2
+      with:
+        message: |
+          There are ${{ steps.link_check.outputs.link_check }} broken links. Check the output of the link check build step for details.
```
Comment (jmelot) [inline .github/workflows/main.yml:91]: Could we display the broken links here to save the effort of opening the logs?

**Review by jmelot — COMMENTED** [review]
I like the way this is headed!

File: .github/workflows/main.yml (hunk 1 above)
Comment (niharikasingh) [reply .github/workflows/main.yml]: Oops, I commented out those steps to speed up the testing process but that's not needed anymore.

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -70,3 +70,27 @@ jobs:
...
         pre-commit run --all-files
+    - name: Build output files
+      run: |
+        cd supply-chain
+        npm run build
+    - name: Check links in built files
+      id: link_check
+      run: |
+        cd supply-chain
+        find public -name "*.js" -exec grep -Eo "(http|https)://[a-zA-Z0-9./?=_%:()-]+" {} \; | sort -u > linklist.txt
```
Comment (jmelot) [inline .github/workflows/main.yml:81]: When I was looking at this again I realized we may need to include '&' to handle query params e.g. match all of something like `https://sciencemap.eto.tech/?china_affiliation_share=50%2C100&NP=50%2C13737&showList=1&cols=cluster_id%2Ccset_extracted_phrase%2Cclass_arts%2Cgrowth_3yr_p_rank%2CNP%2Cchina_affiliation_share%2Cindustry_share`

I wonder if in our case we could just get away with something like (I haven't tested this in bash, just python) `r"[\"\('](http|https://[^\"']+)[\"'\)]"` since links will appear surrounded by those characters in most or all cases we care about - but if you think we're good to go with the current regex, ok to keep it!

Comment (niharikasingh) [reply .github/workflows/main.yml]: Good idea. I think a broader regex will be useful here. I've edited your second regex slightly and added it in.
